feat: implementation of a pid dp controller

Talhanc opened the pull request:
DP controller with quaternion representation used.
The controller is tuned and works in the simulator

Added in docstring for the functions in .hpp file
ROS2 functions are defined separate from the normal functions ( cough Christopher asked for this cough )

Source: Trust me bro

Andeshog commented on guidance/reference_filter_dp/package.xml:
```diff
@@ -0,0 +1,23 @@
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="andeshog@stud.ntnu.no">andeshog</maintainer>
+  <license>TODO: License declaration</license>
```
add license

Andeshog commented on guidance/reference_filter_dp/src/reference_filter_ros.cpp:
```diff
@@ -0,0 +1,242 @@
+        rclcpp::QoSInitialization(qos_profile.history, 1), qos_profile);
+
+    reference_pub_ = this->create_publisher<vortex_msgs::msg::ReferenceFilter>(
+        "/dp/reference", 10);
```
have the topics in the config file

Andeshog commented on guidance/reference_filter_dp/src/reference_filter_ros.cpp:
```diff
@@ -0,0 +1,242 @@
+                  std::placeholders::_1));
+
+    state_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+        "/nucleus/odom", 10,
```
same as above

Andeshog commented on mission/joystick_interface_auv/config/joystick_guidance_params.yaml:
```diff
@@ -0,0 +1,8 @@
+joystick_interface_auv:
+  ros__paramteres:
+    surge_scale_ref_param: 300
+    sway_scale_ref_param: 300
+    heave_scale_ref_param: 300
+    roll_scale_ref_param: 300
+    pitch_scale_ref_param: 300
+    yaw_scale_ref_param: 300
```
why is this here?

Talhanc commented on control/pid_controller_dp/package.xml:
```diff
@@ -0,0 +1,22 @@
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="andeshog@stud.ntnu.no">andeshog</maintainer>
+  <license>TODO: License declaration</license>
```
nice

Talhanc commented on control/pid_controller_dp/package.xml:
```diff
@@ -0,0 +1,22 @@
+<package format="3">
+  <name>pid_controller_dp</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
```
nice

Talhanc commented on control/pid_controller_dp/include/pid_controller_dp/pid_controller.hpp:
```diff
@@ -0,0 +1,47 @@
+    Eigen::Matrix6d Ki_;
+    Eigen::Matrix6d Kd_;
+    Eigen::Vector7d integral_;
+    double dt;
```
fixed

Talhanc commented on guidance/reference_filter_dp/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
```
Why did you add two?

Andeshog commented on control/pid_controller_dp/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+  vortex_msgs
+)
+
+install(
+  DIRECTORY include/
+  DESTINATION include
+)
```
remove this part, as it is already done further down in the file

Andeshog commented on control/pid_controller_dp/include/pid_controller_dp/pid_controller.hpp:
```diff
@@ -0,0 +1,47 @@
+    // @param nu: 6D vector containing the vehicle velocity [u, v, w, p, q, r]
+    // @param eta_dot_d: 7D vector containing the desired vehicle velocity [u,
+    // v, w, p, q, r]
+    // @return 6D vector containing the control input tau [u, v, w, p, q, r]
```
That is not how tau looks.

Talhanc replied:
[Fx, Fy, Fz, torque_x, torque_y, torque_z], Dette bedre?

Talhanc replied:
[X,Y,Z, K, M, N]

Andeshog commented on control/pid_controller_dp/include/pid_controller_dp/pid_controller.hpp:
```diff
@@ -0,0 +1,47 @@
+    explicit PIDController();
+
+    // @brief Calculate the control input tau
+    // @param eta: 7D vector containing the vehicle pose [x, y, z, w, x, y, z]
+    // @param eta_d: 7D vector containing the desired vehicle pose [x, y, z, w,
+    // x, y, z]
```
consider using a different notation for quat, since having two sets of xyz in the same vector can be misleading/confusing.

chrstrom replied:
Also see my comment relating to quaternion operations, `Eigen::Quaterniond` exists for a good reason :fire:

Andeshog commented on control/pid_controller_dp/launch/pid_controller_dp.launch.py:
```diff
@@ -0,0 +1,22 @@
+        executable="pid_controller_node",
+        name="pid_controller_node",
+        parameters=[
+            path.join(
+                get_package_share_directory("pid_controller_dp"),
+                "config",
+                "pid_params.yaml",
+            )
+        ],
```
consider defining this path as a variable

`parameters=[pid_params],`

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_ros.hpp:
```diff
@@ -0,0 +1,98 @@
+
+    nav_msgs::msg::Odometry current_state_;
+
+    // The state vector with 18 elements, [eta, nu, nu_dot]
+    // nu and eta are 6 degrees of freedom (position and orientation in 3D
+    // space)
```
x is [eta, eta_dot, eta_dot_dot] (ref. page 337 in Fossen, 2021

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_utils.hpp:
```diff
@@ -0,0 +1,26 @@
+#include <reference_filter_dp/eigen_typedefs.hpp>
+
+// @brief Calculate the rotation matrix from the Euler angles
+// @param eta The Euler angles
```
eta is more than the euler angles in this case

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_utils.hpp:
```diff
@@ -0,0 +1,26 @@
+Matrix3d calculate_R(const Vector6d& eta);
+
+// @brief Calculate the transformation matrix from the Euler angles
+// @param eta The Euler angles
```
same as above

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_utils.hpp:
```diff
@@ -0,0 +1,26 @@
+Matrix3d calculate_T(const Vector6d& eta);
+
+// @brief Calculate the Jacobian matrix from the Euler angles
+// @param eta The Euler angles
```
same as above

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_utils.hpp:
```diff
@@ -0,0 +1,26 @@
+// @return The Jacobian matrix
+Matrix6d calculate_J(const Vector6d& eta);
+
+// @brief Calculate the shortest signed angle
```
smallest signed angle

Andeshog commented on guidance/reference_filter_dp/include/reference_filter_dp/reference_filter_utils.hpp:
```diff
@@ -0,0 +1,26 @@
+
+// @brief Calculate the shortest signed angle
+// @param angle The angle
+// @return The shortest signed angle
```
same as above

Andeshog commented on guidance/reference_filter_dp/launch/reference_filter.launch.py:
```diff
@@ -0,0 +1,22 @@
+        package='reference_filter_dp',
+        executable='reference_filter_node',
+        name='reference_filter_node',
+        parameters=[
+            os.path.join(
+                get_package_share_directory('reference_filter_dp'),
+                'config',
+                'reference_filter_params.yaml',
+            )
+        ],
```
again, consider making the config path a variable

chrstrom reviewed (changes requested):
First review pass here, I think I'll stop by the office tomorrow if you have any other questions 🫡 (There are also a few choices here relating to data structures and how the code is laid out that I think it's easier to discuss in-person)

Talhanc commented on control/pid_controller_dp/src/pid_controller_utils.cpp:
```diff
@@ -0,0 +1,172 @@
+    error_norm.tail<4>() = q_error;
+
+    Eigen::Vector7d integral_anti_windup = integral + (error_nom * dt);
+
```
i fix this yesterday :innocent:

chrstrom commented on guidance/reference_filter_dp/src/reference_filter_utils.cpp:
```diff
@@ -0,0 +1,60 @@
+
+    if (ctheta == 0) {
+        throw std::runtime_error("Division by zero in calculate_T.");
+    }
```
Forgot to add a comment here regarding what we talked about at the office, but TLDR don't throw just to catch it again. Make sure you either throw and catch it at an appropriate stage (i.e. at the caller or at the very top, then don't publish anything), or crash and die immediately

Talhanc commented:
@kluge7 Bot-en feila man, kan du se på det? builder i min

Andeshog commented on guidance/reference_filter_dp/config/reference_filter_params.yaml:
```diff
@@ -0,0 +1,7 @@
+reference_filter_node:
+  ros__parameters:
+    reference_filter_topic: /aruco_board_pose_camera
+    dp_reference_topic: /dp/reference
+    nucleus_odom_topic: /orca/odom
```
update name for odom and add twist and pose topics

Talhanc commented on control/pid_controller_dp/include/pid_controller_dp/pid_controller_utils.hpp:
```diff
@@ -0,0 +1,87 @@
+// @return 3x3 rotation matrix
+// REF: Handbook of Marine Craft Hydrodynamics and Motion Control, Fossen 2021
+// p.34 eq: 2.72
+types::Matrix3d calculate_R_quat(const types::Eta& eta);
+
+// @brief Calculate the transformation matrix from a quaternion
+// @param q: Quaternion represented as a 4D vector [w, x, y, z]
+// @return 4x3 transformation matrix
+// REF: Handbook of Marine Craft Hydrodynamics and Motion Control, Fossen 2021
+// p.35 eq: 2.78
+types::Matrix4x3d calculate_T_quat(const types::Eta& eta);
+
+// @brief Calculate the Jacobian matrix
+// @param eta: 7D vector containing the vehicle pose [x, y, z, w, x, y, z]
+// @return 7x6 Jacobian matrix
+// REF: Handbook of Marine Craft Hydrodynamics and Motion Control, Fossen 2021
+// p.36 eq: 2.83
+types::J_transformation calculate_J(const types::Eta& eta);
```
I dont like this idea, okey as it is :)

Talhanc commented:
This pull request introduces a new adaptive backstepping controller for dynamic positioning (DP) in a ROS2 package. The changes include adding CMake and package configuration files, defining the controller's parameters, implementing the controller logic, and setting up ROS2 nodes and launch files.

Key changes include:

### Project Configuration:
* [`control/dp_adapt_backs_controller/CMakeLists.txt`](diffhunk://#diff-961596f0b7f904ea3a30789927979c01ea36c238b1bfb91f95dd773bc7b04331R1-R48): Added CMake configuration for building the `dp_adapt_backs_controller` project, including setting the C++ standard, adding dependencies, and defining the executable.
* [`control/dp_adapt_backs_controller/package.xml`](diffhunk://#diff-c76b9e81e21a78795d5fe98446bfc3f56b8b2320bc33577687700d4292097e6cR1-R22): Added package configuration file specifying package dependencies and metadata.

### Controller Implementation:
* [`control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/dp_adapt_backs_controller.hpp`](diffhunk://#diff-0f8d84832952d6dbf4062c775c234e45761f490c32fdca9a42df78d4eebf00f3R1-R71): Defined the `DPAdaptBacksController` class with methods to calculate control input and set various parameters.
* [`control/dp_adapt_backs_controller/src/dp_adapt_backs_controller.cpp`](diffhunk://#diff-b2d623065ce3f5b8cff926341fce62aa34d464716b4966f7f047beea54bbe4b9R1-R103): Implemented the `DPAdaptBacksController` class methods, including the main control logic.

### ROS2 Node and Launch:
* [`control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/dp_adapt_backs_controller_ros.hpp`](diffhunk://#diff-c4afb66cfa37b0fcae7404baea7de55945f538aa31723b491d1172f0bdb8336bR1-R87): Defined the `DPAdaptBacksControllerNode` class for integrating the controller with ROS2, including callback functions for various topics.
* [`control/dp_adapt_backs_controller/launch/dp_adapt_backs_controller.launch.py`](diffhunk://#diff-5b3d7a532d293bc1e1f74d14dde7230a4f1c610768b26649784c00fb9bfb8b8bR1-R24): Added a launch file to start the controller node with specified parameters.

### Additional Utilities and Configuration:
* [`control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/dp_adapt_backs_controller_utils.hpp`](diffhunk://#diff-04549a8d984fdee8d251ae7b5127a5c7c1876add476fa704dec828429d52d726R1-R95): Added utility functions for mathematical operations used in the controller.
* [`control/dp_adapt_backs_controller/config/adapt_params.yaml`](diffhunk://#diff-a71445b73687d2c8cf123a94941adafedadafa223703f7f49afcf9c4b65c3fabR1-R17): Added configuration file for controller parameters.
* [`control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/typedefs.hpp`](diffhunk://#diff-461700671c323ab3121f39812e4f5ab759b8b0ed7e76a58770ee6a5d4d740d23R1-R57): Defined type aliases and structs for the controller.
source: Copilot

Andeshog commented on control/dp_adapt_backs_controller/src/dp_adapt_backs_controller_utils.cpp:
```diff
@@ -0,0 +1,177 @@
+    return euler;
+}
+
+dp_types::Matrix3d calculate_R(const dp_types::Eta& eta) {
+    double cos_phi = std::cos(eta.ori.x());
+    double sin_phi = std::sin(eta.ori.x());
+    double cos_theta = std::cos(eta.ori.y());
+    double sin_theta = std::sin(eta.ori.y());
+    double cos_psi = std::cos(eta.ori.z());
+    double sin_psi = std::sin(eta.ori.z());
+
+    dp_types::Matrix3d R;
+    R << cos_psi * cos_theta,
+        -sin_psi * cos_phi + cos_psi * sin_theta * sin_phi,
+        sin_psi * sin_phi + cos_psi * cos_phi * sin_theta, sin_psi * cos_theta,
+        cos_psi * cos_phi + sin_phi * cos_theta * sin_psi,
+        -cos_psi * sin_phi + sin_psi * cos_phi * sin_theta, -sin_theta,
+        cos_theta * sin_phi, cos_theta * cos_phi;
+    return R;
```
[Comment from Christopher](https://github.com/vortexntnu/vortex-auv/pull/499#discussion_r1835144637)

Andeshog commented on control/dp_adapt_backs_controller/src/dp_adapt_backs_controller_utils.cpp:
```diff
@@ -0,0 +1,177 @@
+    double cos_theta = std::cos(eta.ori.y());
+    double tan_theta = std::tan(eta.ori.y());
+
+    dp_types::Matrix3d T;
+    T << 1, sin_phi * tan_theta, cos_phi * tan_theta, 0, cos_phi, -sin_phi, 0,
+        sin_phi / cos_theta, cos_phi / cos_theta;
```
maybe format this better so that it is easier to read

Talhanc replied:
Is this better? Problem is that the DevOps formating auto sets it to that.
```c++
    dp_types::Matrix3d T;
    T(0, 0) = 1;
    T(0, 1) = sin_phi * tan_theta;
    T(0, 2) = cos_phi * tan_theta;
    T(1, 0) = 0;
    T(1, 1) = cos_phi;
    T(1, 2) = -sin_phi;
    T(2, 0) = 0;
    T(2, 1) = sin_phi / cos_theta;
    T(2, 2) = cos_phi / cos_theta;
```

Andeshog replied:
Ah I see. I think that is better yeah.

Andeshog commented on control/dp_adapt_backs_controller/src/dp_adapt_backs_controller_utils.cpp:
```diff
@@ -0,0 +1,177 @@
+
+dp_types::Matrix3d calculate_R_dot(const dp_types::Eta& eta,
+                                   const dp_types::Nu& nu) {
+    dp_types::Matrix3d R = calculate_R(eta);
+    dp_types::Matrix3d S;
+    S << 0, -1, 0, 1, 0, 0, 0, 0, 0;
```
why not use your method [skew_symmetric](https://github.com/vortexntnu/vortex-auv/blob/434-task-dp-controller/control/dp_adapt_backs_controller/src/dp_adapt_backs_controller_utils.cpp#L19-L25) here?

Andeshog commented on control/pid_controller_dp_euler/include/pid_controller_dp_euler/typedefs.hpp:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    // @brief Make the rotation matrix according to eq. 2.31 in Fossen, 2021
+    Matrix3d as_rotation_matrix() const {
+        double cphi = cos(roll);
+        double sphi = sin(roll);
+        double ctheta = cos(pitch);
+        double stheta = sin(pitch);
+        double cpsi = cos(yaw);
+        double spsi = sin(yaw);
+
+        double r11 = cpsi * ctheta;
+        double r12 = cpsi * stheta * sphi - spsi * cphi;
+        double r13 = cpsi * stheta * cphi + spsi * sphi;
+        double r21 = spsi * ctheta;
+        double r22 = spsi * stheta * sphi + cpsi * cphi;
+        double r23 = spsi * stheta * cphi - cpsi * sphi;
+        double r31 = -stheta;
+        double r32 = ctheta * sphi;
+        double r33 = ctheta * cphi;
+
+        Matrix3d rotation_matrix;
+        rotation_matrix << r11, r12, r13, r21, r22, r23, r31, r32, r33;
+        return rotation_matrix;
```
[Same thing here](https://github.com/vortexntnu/vortex-auv/pull/499#discussion_r1835144637)

Andeshog commented on control/pid_controller_dp_euler/include/pid_controller_dp_euler/typedefs.hpp:
```diff
@@ -0,0 +1,121 @@
+        double t32 = sphi / ctheta;
+        double t33 = cphi / ctheta;
+
+        Matrix3d transformation_matrix;
+
+        transformation_matrix << t11, t12, t13, t21, t22, t23, t31, t32, t33;
```
And here, regarding formatting so it is easier to read

Andeshog commented on control/pid_controller_dp_euler/src/pid_controller_ros.cpp:
```diff
@@ -0,0 +1,139 @@
+    this->declare_parameter("software_operation_mode_topic",
+                            "/softwareOperationMode");
+
+    std::string nucleus_odom_topic =
+        this->get_parameter("odom_topic").as_string();
+    std::string dp_reference_topic =
+        this->get_parameter("reference_topic").as_string();
```
could you also fix the variable names here to "odom_topic" and "reference_topic"?

Andeshog replied:
on a second note, we dont need odom for now actually. Instead add pose and twist as done [here](https://github.com/vortexntnu/vortex-auv/blob/434-task-dp-controller/guidance/reference_filter_dp/src/reference_filter_ros.cpp)

Andeshog commented on control/pid_controller_dp_euler/src/pid_controller_ros.cpp:
```diff
@@ -0,0 +1,139 @@
+    software_mode_ = msg->data;
+}
+
+void PIDControllerNode::odometry_callback(
+    const nav_msgs::msg::Odometry::SharedPtr msg) {
+    eta_.x = msg->pose.pose.position.x;
+    eta_.y = msg->pose.pose.position.y;
+    eta_.z = msg->pose.pose.position.z;
+
+    tf2::Quaternion quat;
+    quat.setX(msg->pose.pose.orientation.x);
+    quat.setY(msg->pose.pose.orientation.y);
+    quat.setZ(msg->pose.pose.orientation.z);
+    quat.setW(msg->pose.pose.orientation.w);
+
+    tf2::Matrix3x3 m(quat);
+    double roll, pitch, yaw;
+    m.getRPY(roll, pitch, yaw);
+
+    eta_.roll = roll;
+    eta_.pitch = pitch;
+    eta_.yaw = yaw;
+
+    nu_.u = msg->twist.twist.linear.x;
+    nu_.v = msg->twist.twist.linear.y;
+    nu_.w = msg->twist.twist.linear.z;
+    nu_.p = msg->twist.twist.angular.x;
+    nu_.q = msg->twist.twist.angular.y;
+    nu_.r = msg->twist.twist.angular.z;
+}
```
split this into pose and twist callback

Andeshog commented on control/pid_controller_dp_euler/src/pid_controller_utils.cpp:
```diff
@@ -0,0 +1,69 @@
+#include <pid_controller_dp_euler/pid_controller_utils.hpp>
+
+Matrix6d float64multiarray_to_diagonal_matrix6d(
+    const std_msgs::msg::Float64MultiArray& msg) {
+    Matrix6d matrix = Matrix6d::Zero();
+
+    if (msg.data.size() != 6) {
+        throw std::runtime_error(
+            "Float64MultiArray message must have exactly 6 elements.");
+    }
+
+    for (size_t i = 0; i < 6; ++i) {
+        matrix(i, i) = msg.data[i];
+    }
+
+    return matrix;
+}
```
this method should not be used anymore so it can be removed

Andeshog commented on guidance/reference_filter_dp/src/reference_filter_ros.cpp:
```diff
@@ -0,0 +1,281 @@
+        goal_handle->publish_feedback(feedback);
+        reference_pub_->publish(feedback_msg);
+
+        if ((x_.head(6) - r_.head(6)).norm() < 0.1) {
+            result->success = true;
+            goal_handle->succeed(result);
+            x_.head(6) = r_.head(6);
+            vortex_msgs::msg::ReferenceFilter feedback_msg =
+                fill_reference_msg();
+            reference_pub_->publish(feedback_msg);
+            RCLCPP_INFO(this->get_logger(), "Goal reached");
+            return;
+        }
```
It would be nice if this was a method actually, called something like goal_reached, so that this block becomes something like
```C++
if (goal_reached()) {
    return;
}

Talhanc replied:
To be honest, i dont know how i would do this

Andeshog replied:
my bad, I was wrong. what you can do though is make this block its own function called something like `check_if_goal_reached()` so that it just becomes becomes 
```C++
check_if_goal_reached()
```

Andeshog commented on guidance/reference_filter_dp/src/reference_filter_utils.cpp:
```diff
@@ -0,0 +1,60 @@
+#include <reference_filter_dp/reference_filter_utils.hpp>
+
+Matrix3d calculate_R(const Vector6d& eta) {
+    double phi = eta(3);
+    double theta = eta(4);
+    double psi = eta(5);
+
+    double cphi = cos(phi);
+    double sphi = sin(phi);
+    double ctheta = cos(theta);
+    double stheta = sin(theta);
+    double cpsi = cos(psi);
+    double spsi = sin(psi);
+
+    Matrix3d rotation_matrix;
+
+    rotation_matrix << cpsi * ctheta, cpsi * stheta * sphi - spsi * cphi,
+        cpsi * stheta * cphi + spsi * sphi, spsi * ctheta,
+        spsi * stheta * sphi + cpsi * cphi, spsi * stheta * cphi - cpsi * sphi,
+        -stheta, ctheta * sphi, ctheta * cphi;
+
+    return rotation_matrix;
+}
```
why was this conversation marked as resolved 😭

Andeshog commented on guidance/reference_filter_dp/src/reference_filter_utils.cpp:
```diff
@@ -0,0 +1,56 @@
+    double stheta = sin(theta);
+
+    // Manually added the transformation matrix from Fossen 2021 p.29 eq: 2.41
+    transformation_matrix << 1, sphi * stheta / ctheta, cphi * stheta / ctheta,
+        0, cphi, -sphi, 0, sphi / ctheta, cphi / ctheta;
```
readability could be improved here by making it look somewhat like a 3x3 matrix

Andeshog reviewed (commented):
+ add README on all 4, although I could write the one about the reference filter

Talhanc commented on control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/dp_adapt_backs_controller.hpp:
```diff
@@ -0,0 +1,71 @@
+
+    // @brief Set the K1 matrix
+    // @param K1: 6D vector containing the K1 matrix
+    void setK1(const dp_types::Vector6d& K1);
+
+    // @brief Set the K2 matrix
+    // @param K2: 6D vector containing the K2 matrix
+    void setK2(const dp_types::Vector6d& K2);
+
+    // @brief Set the r_b_bg vector
+    // @param r_b_bg: 3D vector containing the r_b_bg vector
+    void setrbg(const dp_types::Vector3d& r_b_bg);
+
+    // @brief Set the adaptive parameter matrix
+    // @param adap_param: 12D vector containing the adaptive parameter matrix
+    void setAdapParam(const dp_types::Vector12d& adap_param);
+
+    // @brief Set the d gain matrix
+    // @param d_gain: 6D vector containing the d gain matrix
+    void setDGain(const dp_types::Vector6d& d_gain);
+
+    // @brief Set the inertia matrix
+    // @param I_b: 3D vector containing the inertia matrix
+    void setInertiaMatrix(const dp_types::Vector3d& I_b);
+
+    // @brief Set the mass inertia matrix
+    // @param m: Mass inertia matrix
+    void setMassInertiaMatrix(const dp_types::Matrix6d& M);
+
+    // @brief Set the mass
+    // @param m: Mass
+    void setm(double m);
+
+    // @brief Set the time step
+    // @param dt: Time step
+    void setTimeStep(double dt);
```
fixed

Talhanc commented on control/dp_adapt_backs_controller/src/dp_adapt_backs_controller_node.cpp:
```diff
@@ -0,0 +1,12 @@
+    rclcpp::init(argc, argv);
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
+                "Started DP Adaptive Backstepping Controller Node");
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
+                "-- Owned and licensed by Talha Nauman Choudhry --");
```
I am HIM

chrstrom replied:
\>"Owned and licensed by Talha Nauman Choudhry"

\>**looks inside**

\>_MIT License Copyright (c) 2020 Vortex NTNU_

chrstrom commented on control/dp_adapt_backs_controller/include/dp_adapt_backs_controller/dp_adapt_backs_controller_ros.hpp:
```diff
@@ -0,0 +1,87 @@
+
+// @brief Class for the DP Adaptive Backstepping controller node
+class DPAdaptBacksControllerNode : public rclcpp::Node {
+   public:
+    explicit DPAdaptBacksControllerNode();
```
Taking no inputs in the constructor and relying on the user-side correctly calling set_* could mess you up at some point. Better to take all required inputs in the constructor (pack them appropriately) and/or use a factory

Talhanc replied:
i fixed this, added a struct with parameters that are paseed into the constructer

Talhanc commented:
@Andeshog and @chrstrom I have done everything asked and fixed the issues. Is it time for merging?

chrstrom reviewed (approved):
You really are HIM 🗣️ 🔥  Approving under the assumption that you've tested in sim / with physical orca

Andeshog reviewed (approved):
LGTM :+1:
